File: 2465-shifting-letters-ii/shifting-letters-ii.py
class Solution:
    def shiftingLetters(self, s: str, shifts: List[List[int]]) -> str:
        diff=[0]*(len(s)+1)
        for start, end, direction in shifts:

            value = 1 if direction == 1 else -1

            diff[start] += value
            diff[end + 1] -= value

        curr=0
        ans=[]
        for i in range(len(s)):
            curr+=diff[i]
            curr%=26
            c=s[i]
            new_char=chr(ord('a')+(ord(c)-ord('a')+curr)%26)
            ans.append(new_char)
        return "".join(ans)

File: 2465-shifting-letters-ii/shifting-letters-ii.java
class Solution {
    public String shiftingLetters(String s, int[][] shifts) {

        int n = s.length();

        int[] diff = new int[n + 1];

        // Difference Array
        for (int[] shift : shifts) {

            int start = shift[0];
            int end = shift[1];
            int direction = shift[2];

            int value = (direction == 1) ? 1 : -1;

            diff[start] += value;
            diff[end + 1] -= value;
        }

        // Prefix Sum + build answer
        StringBuilder ans = new StringBuilder();

        int currentShift = 0;

        for (int i = 0; i < n; i++) {

            currentShift += diff[i];

            currentShift %= 26;

            if (currentShift < 0) {
                currentShift += 26;
            }

            char c = s.charAt(i);

            char newChar =
                (char) ('a' + (c - 'a' + currentShift) % 26);

            ans.append(newChar);
        }

        return ans.toString();
    }
}